### DISCUSSION TITLE
How to wait for connection

### DISCUSSION THREAD

**Discussion by atauf** [PhilipsHue/flutter_reactive_ble#321] [Q&A]
This question slightly piggy backs #319, in that, I believe it's only a noticeable issue because the advertisement rate is so slow.

I'd like to know how to implement a wait in the program until a connection is actually made. I'm using writeToCharacteristicWithoutResponse because the devices I'm connecting to don't give a response.

I've wrapped the writeToCharacteristicWithoutResponse function with a function that will produce a response (though I'm not sure if it's done the best way or not), but either way, I've added a response.

I basically have a loop that iterates through a list of device ids and is supposed to connect, load some parameters via multiple writeToCharacteristicWithoutResponse calls, and then disconnects.

code: (note, this is in the onPressed section of an IconButton)

```
for (String deviceId in deviceIdList) {
                try {
                  await Provider.of<BleDeviceConnector>(context, listen: false)
                      .connect(deviceId)
                      .then((value) => loadStuff(context))
                      .then((value) =>
                      Provider.of<BleDeviceConnector>(context, listen: false)
                          .disconnect(deviceId));
                } catch (e) {
                  print(e);
                }
```

The problem that I'm running into is that by the time the first device actually connects, the loop has already moved through all the device ids in the list.

I'd like to know how to actually make it wait for each process to complete.

NOTE: I'm using async/await for all of these futures

Other Mentions: I'm starting the loop from a scanning and/or disconnected state (that's why I connect and disconnect for each).

Thanks for any help, and let me know if I can provide any other info.

**Comment by remonh87** [comment]
@atauf  the connectToDevice provides a stream of connection updates. You can decide to listen to this Stream and only execute operations  when you get the `connected` event. See also: https://github.com/PhilipsHue/flutter_reactive_ble/blob/master/example/lib/src/ble/ble_device_connector.dart

**Reply by atauf** [reply]
@remonh87 Sounds good. Just to clarify, are you suggesting
1) adding to the actual .listen( (update) {...} ) functionality that already exists, or
2) just calling .connectToDevice on a separate occasion and handling it from there.

I'm currently doing 1), but I'm thinking I might still have this same issue. It'll take care of doing something once connected, but then I need to make sure to wait until I write to all the desired characteristics prior to disconnecting. One solution I thought of was to throw an exception when all of the writing to characteristics part is done, and then do the disconnect in the onError section. Not sure how elegant that is, but thought I'd see what you think. Thanks!

**Comment by atauf** [comment]
@remonh87 Is there any reason why I shouldn't be able to loop through a list of deviceIds and successfully execute .disconnect(deviceId) on each of them? I'm currently trying to do it and it only disconnects from one of them. It won't disconnect from the rest.

**Reply by remonh87** [reply]
you can create a class that will hold all subscriptions and then you can loop through them and cancel them one by one in order to disconnect

**Reply by atauf** [reply]
@remonh87 so would I have to add the subscriptions to this class when I create them (currently doing in a loop)? I don't know quite how to do that given the example app's ble_device_connector.dart. I need another class?

Sorry for all the questions, but this just hasn't been easy to do for me and I've been trying to get it done for a few days now.

**Reply by atauf** [reply]
@remonh87 or do you mean a list? I'm just confused as to what a class will do or how to construct it.

**Reply by atauf** [reply]
@remonh87 Here is an update of what the code looks like:

For loop to connect and load parameters:
```
            for (String deviceId in deviceIdList) {
              try {
                await Provider.of<BleDeviceConnector>(context, listen: false)
                    .connectAndLoad(
                        deviceId: deviceId, custom: custom, index: index);
              } catch (e) {
                print(e);
              }
            }
```

Connect and Load function (inside of ble_device_connector.dart):
  ```
Future<void> connectAndLoad(
      {required String deviceId,
      required CustomModel custom,
      required int index}) async {
    _connection = _ble.connectToDevice(id: deviceId).listen((update) async {
      _deviceConnectionController.add(update);
      if (update.connectionState == DeviceConnectionState.connected) {
        await custom.loadCustomStuff();
      }
    }, onError: (e) {
      print('Connecting to device $deviceId resulted in error $e');
    });
  }
```
